### ISSUE TITLE
Pass rpmbuild arguments into Docker

### ISSUE THREAD

**Issue by hanej** [alanfranz/docker-rpm-builder#22]
I would like to override or add new macros that the spec file is using to rpmbuild at runtime.  Typically this is done with the `-D, --define='MACRO EXPR'` argument.  Could you add an environment variable that takes a list of arguments and pass them through to rpmbuild?

Something like this:
`export DRB_RPMBUILD_ARGS="--define dist=el7 --define author='My Name' --define release=32"`

**Comment by alanfranz** [comment]
Hello Jason,
why can't you use the spectemplate mechanism which is provided by docker-rpm-builder? It was designed for such exact purpose.

If you've got values that should be defined per-distribution, on the contrary, I'd suggest you add macros to  /etc/rpm/macros.something in your build image.

**Comment by hanej** [comment]
The use case I'm trying to solve using spec files that come with external projects without having to go through and modify them before building.  I also don't want to keep a static version of the spectemplate which may be out of sync with what comes with the package.  You're right, I could use the spectemplates for projects which I control the spec file.

For overriding macros I didn't want to have to maintain a separate image for each distribution.  I'm using your images (which are very good by the way) and that is working well.  Since docker-rpm-builder grabs the spec file and source from the directory I suppose it could also suck in a macros file as well and drop it in $HOME/.rpmmacros.  That would solve my problem as well.

**Comment by alanfranz** [comment]
I understand. I'm usually against putting things with too a narrow use case into my software, but since you can actually forward arbitrary options from docker-rpm-builder to the docker daemon, I'll try making a change to see whether I can actually let you set RPMBUILD_OPTIONS from the command line (quite similar to what you're doing in your PR, but without the explicit addition in the command).

**Comment by alanfranz** [comment]
I have chosen a different path. Now you've got an extra environment variable you can leverage via --env or --env-file, but I won't provide explicit support via command line, since I think the use case is a bit narrow.

**Comment by alanfranz** [comment]
I think I'll do a stable release this or next week.

**Comment by alanfranz** [comment]
Version 1.27 is out, including the RPMBUILD_EXTRA_OPTIONS support.

**Comment by dansou901** [comment]
Hi Alan,
nice to see an option for rpmbuild args. There is not only the use case described above. If you want to build a new software collection for RHEL like systems and don't have the %sclname-build package for that scl installed in the container, you need to pass --define scl %sclname to rpmbuild. I'm not sure if this is possible via spectemplates. Please also refer to https://www.softwarecollections.org/en/docs/guide/#chap-Packaging_Software_Collections for more information. So, thanks for adding that!
Btw, I also bumped the version number of the AUR package.
Regards,
Daniel

**Comment by hanej** [comment]
Building SCL packages was actually the reason that I started using this project

**Comment by dansou901** [comment]
Do you build your SCLs via --define scl name? If yes, can you give me an example? When I try to use --define scl sclo-git25 for instance, I get: 
`error: Macro %scl has empty body`
P.S.: Sorry for OT

**Comment by dansou901** [comment]
Interesting... While building inside a docker container with rpmbuild -ba --define 'scl sclo-git25' the build works perfectly. But when using docker-rpm-builder with the env variables above, I always get macro %scl has empty body. I assume that special chars don't get escaped/evaluated correctly. Can you have a look at this again, Alan? Thanks in advance!
Regards,
Daniel

**Comment by hanej** [comment]
I haven't tried the new version yet.  I'm still on my forked version.  I should have some time to test the new one out next week.

**Comment by dansou901** [comment]
Maybe this is a side effect of https://github.com/docker/docker/issues/12997?

**Comment by alanfranz** [comment]
@dansou901 I'll gladly look at it if you provide me with a practical example of what you're building and the command line you're using!

**Comment by alanfranz** [comment]
Also, if you run with "--verbose" you should be able to get the full command output and check whether there's an issue with the way the arguments are passed.

**Comment by dansou901** [comment]
I'm currently working on building tomcat scls so that you can install different tomcat versions on one server hosting different webapps. I'm not done with the spec file yet, but I thought it would be good to have a working build environment first, so I tried building with the sclo-git25 collection. I downloaded the sources from https://github.com/sclorg-distgit/git/tree/sig-sclo7-sclo-git25-sclo, navigated to that directory (you'll have to move the sources from the sources folder to the parent directory to be able to build with docker-rpm-builder) and did a `docker-rpm-builder dir myimage . /home/daniel/rpm -- --env=RPMBUILD_OPTIONS="define 'scl sclo-git25'"`. Pay attention to the quotes around scl sclo-git25, without them I get the error described above. Those quotes won't get interpreted correctly. So basically I need a way to define macros safely via an option for docker-rpm-builder to be able to build scls.

**Comment by alanfranz** [comment]
@dansou901 I probably need to add this data the doc. You must pass RPMBUILD_EXTRA_OPTIONS , not RPMBUILD_OPTIONS - the latter is for internal use only.

**Comment by alanfranz** [comment]
And you must properly format the option, it's not an extra define, just an option - it would be something like

```
--env=RPMBUILD_EXTRA_OPTIONS="--define 'scl sclo-git25'"
```

**Comment by dansou901** [comment]
Thanks, will try out and let you know.

**Comment by dansou901** [comment]
When I format the option the way you did, it is evaluated like this:
`rpmbuild --define ''\''scl' 'sclo-git25'\''' -ba /root/rpmbuild/SPECS/git.spec`
So there must be something wrong how you evaluate quotes...

**Comment by alanfranz** [comment]
Totally right. Additional options should not be escaped, it's up to the caller to pass them properly, since they got straight into docker. I'm committing this right now, 1.28 will be out around 1pm CEST, i suppose. Test that once it's out.

**Comment by dansou901** [comment]
Thanks for the fix, I will push the AUR package to the new version as well as soon as it's out and test then.

**Comment by alanfranz** [comment]
1.28 is out. try!

**Comment by dansou901** [comment]
Hi, I don't see the fix in v1 yet, so I won't push the AUR package because I've sourced v1 there...

**Comment by dansou901** [comment]
Nevermind, don't push it yet, it doesn't work. If I try the command line above, the quotes around --define 'scl sclo-git25' won't get recognized at all...

**Comment by dansou901** [comment]
After `docker-rpm-builder dir dansou901/centos-scl . /home/daniel/rpm/sclo7-sclo-git25-sclo/ -- --env=RPMBUILD_EXTRA_OPTIONS="define 'scl sclo-git25'"` the terminal gives me something like this:

```
2016-05-06T16:44:55+0200 INFO [drb.commands.dir] Now building project from /home/daniel/sclorg-distgit-git/git on image dansou901/centos-scl
2016-05-06T16:44:55+0200 ERROR [UserExceptionTransformer] Full catched exception:
Traceback (most recent call last):
  File "/opt/docker-rpm-builder/lib/python2.7/site-packages/drb/commands/dir.py", line 166, in dir
    docker.do_run()
  File "/opt/docker-rpm-builder/lib/python2.7/site-packages/drb/docker.py", line 75, in do_run
    return self._run().strip()
  File "/opt/docker-rpm-builder/lib/python2.7/site-packages/drb/docker.py", line 95, in _run
    raise SpawnedProcessError(retcode, fullcmd, output=output, error=error)
SpawnedProcessError: Command '/usr/bin/docker run --tmpfs=/tmp --rm --env=RPMBUILD_EXTRA_OPTIONS=define 'scl sclo-git25' --volume=/home/daniel/sclorg-distgit-git/git/git.spec:/root/rpmbuild/SPECS/git.spec:ro --volume=/opt/docker-rpm-builder/lib/python2.7/site-packages/drb/dockerscripts:/dockerscripts:ro --volume=/home/daniel/sclorg-distgit-git/git:/root/rpmbuild/SOURCES:ro --volume=/home/daniel/rpm/sclo7-sclo-git25-sclo:/root/rpmbuild/RPMS --workdir=/dockerscripts --env=ENABLE_SOURCE_OVERLAY=1 --env=CALLING_UID=1000 --env=CALLING_GID=1000 --env=BASH_ON_FAIL='' dansou901/centos-scl ./rpmbuild-dir-in-docker.sh' returned non-zero exit status 125:

/usr/bin/docker: Error parsing reference: "scl sclo-git25" is not a valid repository/tag.
See '/usr/bin/docker run --help'.


Error: docker run error
Command '/usr/bin/docker run --tmpfs=/tmp --rm --env=RPMBUILD_EXTRA_OPTIONS=define 'scl sclo-git25' --volume=/home/daniel/sclorg-distgit-git/git/git.spec:/root/rpmbuild/SPECS/git.spec:ro --volume=/opt/docker-rpm-builder/lib/python2.7/site-packages/drb/dockerscripts:/dockerscripts:ro --volume=/home/daniel/sclorg-distgit-git/git:/root/rpmbuild/SOURCES:ro --volume=/home/daniel/rpm/sclo7-sclo-git25-sclo:/root/rpmbuild/RPMS --workdir=/dockerscripts --env=ENABLE_SOURCE_OVERLAY=1 --env=CALLING_UID=1000 --env=CALLING_GID=1000 --env=BASH_ON_FAIL='' dansou901/centos-scl ./rpmbuild-dir-in-docker.sh' returned non-zero exit status 125:

/usr/bin/docker: Error parsing reference: "scl sclo-git25" is not a valid repository/tag.
See '/usr/bin/docker run --help'.



Build error. See the log above`
```

**Comment by dansou901** [comment]
Got that, if I escape the quotes around the --define option, the build starts, but escaping the macro like  \'scl sclo-git25\' itself produces again
`rpmbuild --define ''\''scl' 'sclo-git25'\''' -ba /root/rpmbuild/SPECS/git.spec`

**Comment by alanfranz** [comment]
I'm taking a further look. It shouldn't behave that way.

**Comment by alanfranz** [comment]
@hanej @dansou901 after years of use, bash and its curious evaluation/escaping approach never ceases to surprise me.

the RPMBUILD_EXTRA_OPTIONS approach is flawed. I'd need to serialize and deserialize an array to properly get that option parsed at that point.

Would it be ok to pass a file? Something like, if /rpmmacros exists, it's copied in ${HOME}/.rpmmacros so it gets read. Otherwise the serialization/deserialization which is needed just gets crazy.

**Comment by hanej** [comment]
That's the reason I had to throw in the evals.  A file works too.  Either is fine with me.  Thank you.

**Comment by dansou901** [comment]
A file is ok with me as well, as long as it works as expected. So the define options need to be placed in a file then?

**Comment by alanfranz** [comment]
Hello, could you please try with the /rpmmacros approach in 1.31 and tell me if that finally works? Doc is within the command help.

**Comment by dansou901** [comment]
I'll try as soon as I get home after work.

**Comment by dansou901** [comment]
Now I get the following error:

```
+ '[' -r /rpmmacros ']'
+ cp /rpmmacros /root/.rpmmacros
cp: omitting directory '/rpmmacros'
```

**Comment by dansou901** [comment]
What I did was I created a file called rpmmacros in the source directory with the content `%scl sclo-git25`. After that I used the command you provided in the docs.

**Comment by alanfranz** [comment]
uhm... but did you specify the path properly? rpmmacros must be a _file_
with the content just like rpmmacros. The content seems OK, but I think you
didn't specify the path properly, so docker created an host directory with
the name you provided, and bind-mounted that dir in the guest.

**Comment by dansou901** [comment]
Sorry, didn't get that at first. Now it's working perfectly. Thanks for your support! I pushed the AUR package to the new version as well.

**Comment by aframachines** [comment]
Great!

If you think the help text is misleading, would you create a pr with an
improvement?

Also, where/how should i provide info about the aur package?
Il 13/mag/2016 05:58 PM, "dansou901" notifications@github.com ha scritto:

**Comment by dansou901** [comment]
PR created. Info about the AUR package should be done in the Installation section on the README file.

**Comment by alanfranz** [comment]
Finally fixed in version 1.31
